plugins/package-managers/python/src/main/kotlin/Poetry.kt

```diff
@@ -64,9 +64,12 @@ class Poetry(
     override fun transformVersion(output: String) = output.substringAfter("version ").removeSuffix(")")
 
     override fun resolveDependencies(definitionFile: File, labels: Map<String, String>): List<ProjectAnalyzerResult> {
-        val result = inspectLockfile(definitionFile)
+        val dependenciesResult = inspectLockfile(definitionFile, includeDevDependencies = false)
+        val devDependenciesResult = inspectLockfile(definitionFile, includeDevDependencies = true)
```

Comment:
`includeDevDependencies = true` sounds as if `devDependenciesResult` contains `dependenciesResult` *and* any dev dependencies. If that's so, shouldn't we substract `dependenciesResult` from `devDependenciesResult` and only use that for a `dev` (not `installDev`) scope?

Reply:
Yes, `devDependenciesResult` includes `dependenciesResult`, but I believe we should not substract, because:

1. The dependencies which end-up in the release artifacts are normally also needed for 
     e.g. test builds
2. It's unclear (to me) how substracting trees should work. We don't do this anywhere, 
    please correct if I'm wrong.
4. It would be inconsistent with e.g. `GoMod` integration and believe also what `.gradlew 
    dependencies` outputs for the respective scopes.

Reply:
@sschuberth - do you agree with above comment, so we can leave as-is?

Reply:
I still don't understand why it should be correct to end up with two scopes, one named "install" and the other named "installDev", where "install" seems to be a subset of "installDev". If I'm not getting anything wrong, this overlap seems counter-intuitive to me, and it's something that we AFAIK do not do for any other package manager.

Reply:
> that we AFAIK do not do for any other package manager.

I doubt this is true. Have you actually read my comment above? I says:

It would be inconsistent with e.g. GoMod integration and believe also what .gradlew  dependencies outputs for the respective scopes.

Reply:
> > that we AFAIK do not do for any other package manager.
> 
> I doubt this is true. Have you actually read my comment above

Yes, I have read it, but it even confused me more. For GoMod there seem to be two scopes created, "main" and "vendor". Simply judging from the names, there seems to be no overlap of dependencies here.

I also didn't get your comment about `./gradlew dependencies`. Gradle lists dependencies per configuration, and it's not combining dependencies of configurations. Note that I'm not talking about cases where a given dependency really happens to be part of multiple configurations / scopes.

Or phrased differently: The "installDev" scope sounds to me as if it should actually be an "install" scope (that we already have) and a "dev" scope. And the only reason why we don't have it like that seems to be because it's harder to implement. That's at least my understanding of the issue.

Reply:
I was just reading up on the [`poetry export` command help](https://python-poetry.org/docs/cli/#export), and could we maybe use `--with dev --without main` to really only get `dev` dependencies?

Reply:
Or yet better ``--only dev`?

Reply:
> Yes, I have read it, but it even confused me more. For GoMod there seem to be two scopes created, "main" and "vendor". Simply judging from the names, there seems to be no inherent overlap of dependencies here.

I believe the scope names do not provide sufficient information. Looking at https://github.com/oss-review-toolkit/ort/blob/1c4c0fc2ec0eca2aea1a6a4ad51dea8a19da98d0/analyzer/src/main/kotlin/managers/GoMod.kt#L126 it shows that `main` is a subset of `vender` in terms of nodes and edges. 

> I also didn't get your comment about `./gradlew dependencies`. Gradle lists dependencies per configuration, and it's not combining dependencies of configurations. Note that I'm not talking about cases where a given dependency really happens to be part of multiple configurations / scopes.

Hm, so are you saying that e.g. `compileClasspath` is not a subset of `testCompileClasspath`?  


> Or phrased differently: The "installDev" scope sounds to me as if it should actually be an "install" scope (that we already have) and a "dev" scope. 

I think it's inappropriate to derive some meaning from these names as they are rather arbitrarily chosen.  I believe it'd be more appropriate to pick better names if they don't fit the implementation we think makes most sens.

> And the only reason why we don't have it like that seems to be because it's harder to implement. That's at least my understanding of the issue.

My understanding may be wrong, but I thought that it would be consistent with what we already have in e.g. GoMod and Gradle. Furthermore, when resolving dev dependencies it could be that constraints coming from the non-dev dependencies are considered. Or phrased differently, could it be that resolving the dev dependcies without considering the main dependencies version constraints could lead to wrong resolved package version?

>I was just reading up on the poetry export command help, and could we maybe use --with dev --without main to really only get dev dependencies?

Interresting idea, that at least would not require one to subtract trees in some not yet clear / specified way and should be simple to implemented. However, I'm not sure whether the results are as good as with the current approach, because of the above mentioned constraints thing. What do you think about that?

Reply:
> Hm, so are you saying that e.g. `compileClasspath` is not a subset of `testCompileClasspath`?

Not necessarily, yes. In practice, that will rarely be the case, but my point is that Gradle allows these to contain distinct dependencies without any overlap. They are not "hard-wired" to be a subset of each other. And if they aren't, our Gradle package manager implementation will still get that right.

Coming back to the Poetry docs, I believe what's called "dependency group" should map to a "scope" in ORT. And as "main" and "dev" (and probably more) are distinct dependency groups, they should also be distinct scopes.

Reply:
> Not necessarily, yes. In practice, that will rarely be the case, but my point is that Gradle allows these to contain distinct dependencies without any overlap. They are not "hard-wired" to be a subset of each other. And if they aren't, our Gradle package manager implementation will still get that right.

Thanks. But is this a theoretical consideration or is it like this in practice? (I'm asking because 
in my memories the scopes I've investigated seemed more like they were subsets)
Also my point about constraints would be interesting here: In practice, do the dev dependencies scopes
in gradle consider the version constraints of the release dependencies scopes? (Or phrasing it differently,
are the contraints of release dependencies scopes usually a subset of the dev dependencies scope) ..I'm asking this explicitly, because if they weren't, the resolution might lead to conflicting versions.  

> Coming back to the Poetry docs, I believe what's called "dependency group" should map to a "scope" in ORT. And as "main" and "dev" (and probably more) are distinct dependency groups, they should also be distinct scopes.

Letting aside the difficulty of the implementation for now, I believe we have to consider my point about constraints 
also for poetry. If we omit the release deps for analyzing the dev deps, then the release deps constraints are not considered. So, resolving the union of the constraints of both scopes can yield a result different from resolving 
dev dependencies constraints and release dependency constraints separately. So, one of the two approaches could yield a result inconsistent with the tooling (poetry). 

Would you agree that we should make correctness in this regard a priority for deciding whether to keep implementation as-is or to change it as suggested?

Reply:
I fear that we again might be having problems understanding each other as we're using terms differently. To me "(dependency) constraints" are somethings like [this](https://docs.gradle.org/current/userguide/dependency_management_terminology.html#sub:terminology_dependency_constraint) in Gradle, i.e. an explicit mean to express additional requirements for transitive dependencies. I'm not aware of Poetry having such a mechanism, hence I believe you're using the term for any implicit "constraint" that results from the same dependency appearing multiple times in different versions / version ranges; i.e. the standard case for general version conflict resolution.

So when you say

> If we omit the release deps for analyzing the dev deps, then the release deps constraints are not considered.

I believe you mean that version conflict resolution might be different for "dev" alone than for a combination of "release" and "dev". But I'd assume that cannot happen as we're completely leaving the dependency resolution to Poetry in all cases; version conflict resolution is nothing that ORT should need to care about. Instead, I'd simply trust `poetry export --only <groupname>` to do the right thing.

Reply:
I've changed it as you suggested, e.g. use `poetry export --only dev`